src/coreComponents/mesh/DomainPartition.cpp

```diff
@@ -183,14 +182,39 @@ void DomainPartition::setupCommunications( bool use_nonblocking )
 
     CommunicationTools::getInstance().findMatchedPartitionBoundaryObjects( nodeManager,
                                                                            m_neighbors );
+  } );
+}
 
-    CommunicationTools::getInstance().setupGhosts( meshLevel, m_neighbors, use_nonblocking );
 
-    faceManager.sortAllFaceNodes( nodeManager, meshLevel.getElemManager() );
-    faceManager.computeGeometry( nodeManager );
+void DomainPartition::setupCommunications( bool use_nonblocking )
+{
+  forMeshBodies( [&]( MeshBody & meshBody )
+  {
+    meshBody.forMeshLevels( [&]( MeshLevel & meshLevel )
+    {
+      if( meshLevel.getName() == MeshBody::groupStructKeys::baseDiscretizationString() )
```

Comment:
It's complicated for me to get the diff, but I can see that there may be cases where nothing will be done for the `meshLevel` (if the `meshLevel.getName()` does not match, or if it's not a shallow copy)?
Is it intended? That sounds like a bit dangerous?

Reply:
Thanks for asking. The statements in the if statement are for base mesh-level. And the statements in elseif (not shallowcopy) are for high-order mesh-levels.

Reply:
Then I would say that there should be a `else` branch which `errors` if none of the previous `if` match.